boa_interner/src/lib.rs

```diff
@@ -119,21 +131,42 @@ impl Interner {
         T: AsRef<str>,
     {
         let string = string.as_ref();
-        Self::get_static(string).or_else(|| self.inner.get(string))
+        Self::get_common(string).or_else(|| self.symbols.get(string).copied())
     }
 
     /// Interns the given string.
     ///
     /// Returns a symbol for resolution into the original string.
     ///
     /// # Panics
+    ///
     /// If the interner already interns the maximum number of strings possible by the chosen symbol type.
     pub fn get_or_intern<T>(&mut self, string: T) -> Sym
     where
         T: AsRef<str>,
     {
         let string = string.as_ref();
-        Self::get_static(string).unwrap_or_else(|| self.inner.get_or_intern(string))
+        if let Some(sym) = self.get(string) {
+            return sym;
+        }
+
+        let capacity = self.head.capacity();
+
+        if capacity < self.head.len() + string.len() {
+            let new_cap = (usize::max(capacity, string.len()) + 1).next_power_of_two();
+            let new_head = String::with_capacity(new_cap);
+            let old_head = std::mem::replace(&mut self.head, new_head);
+            self.full.push(old_head);
+        }
```

Comment:
I think it would be nice to have a comment explaining why this is done and how the pointers into the old `head` are still valid even after pushing it to `full`. Probably that comment could also be on the fields of `Interner` directly.

Reply:
Changed a bit the implementation to better reflect the unsafe parts, and documented some other parts with what you've mentioned